Implement isMinTTY

RyanGlScott opened the pull request:
Fixes #62. This adapts code written [here](https://phabricator.haskell.org/D2809) (by Phil Ruffwind) and [here](https://github.com/git-for-windows/git/blob/4af28e2fb4af1069c81a3dcfab0332498e8424de/compat/winansi.c#L624-L649) (from `git-for-windows`) to provide a backwards-compatible way to tell if you're running in a MinTTY console on Windows.

I have some concerns that I'll leave as inline comments.

RyanGlScott commented on System/Win32/MinTTY.hsc:
```diff
@@ -0,0 +1,207 @@
+-- Returns 'False' otherwise.
+isMinTTY :: IO Bool
+isMinTTY = do
+    h <- getStdHandle sTD_ERROR_HANDLE
```
This currently just hardwires in `sTD_ERROR_HANDLE`. Do you think that's acceptable, or should we provide a version of `isMinTTY` that accepts any `HANDLE`?

RyanGlScott commented on System/Win32/MinTTY.hsc:
```diff
@@ -0,0 +1,207 @@
+    getFileInformationByHandleEx h fILE_NAME_INFO buf (fromIntegral bufSize)
+    len :: DWORD <- peek buf
+    let len' = fromIntegral len `div` sizeOfWchar
+    peekCWStringLen (buf `plusPtr` sizeOfDWORD, min len' mAX_PATH)
```
This (and `ntQueryObjectNameInformation` below) do a lot of memory manipulation which, as far as I can see, making some assumptions about how the memory being written to `buf` is laid out. We have to do this since the structs that we need (`FILE_NAME_INFO` and `OBJECT_NAME_INFORMATION`, respectively) don't appear to be available on the version of MinGW-w64 that ship with GHC (8.0.1, at least).

One alternative would be conditionally defining these structs in a header file (if they aren't already defined), and rely on the FFI to marshal them via `Storable` instances.

RyanGlScott commented on System/Win32/MinTTY.hsc:
```diff
@@ -0,0 +1,207 @@
+  mk_RtlGetVersion :: FunPtr F_RtlGetVersion -> F_RtlGetVersion
+
+type ULONG    = #type ULONG
+type NTSTATUS = #type NTSTATUS
```
Perhaps `ULONG` and `NTSTATUS` belong in `System.Win32.Types`? Not sure.

RyanGlScott commented on System/Win32/MinTTY.hsc:
```diff
@@ -0,0 +1,207 @@
+  , dwBuildNumber       :: ULONG
+  , dwPlatformId        :: ULONG
+  , szCSDVersion        :: CWString
+  } deriving Show
```
The ability to query your Windows version might be handy as its own feature. Thoughts?

Mistuke replied:
It is, but I'm currently in the process of merging in `Win32-Extras`. Which already provides these :)

RyanGlScott commented:
Thanks for the review, @Mistuke. I've implemented your suggestions:

1. I removed the unnecessary Vista version check.
2. I moved `ULONG` to `System.Win32.Types` (and made a note of it in the changelog)
3. I defined `WINVER` and `_WIN32_WINNT` to be `0x0600` so as to be able to use the `FILE_NAME_INFO`, `OBJECT_NAME_INFORMATION`, and `UNICODE_STRING` structs. Unforuntately, this now causes a rather ugly redefinition warning whenever you compile `Win32`:

   ```
   Preprocessing library Win32-2.4.0.0...
   MinTTY.hsc:48:0: warning: "_WIN32_WINNT" redefined
   In file included from C:/Users/RyanGlScott/Software/ghc-8.0.1/mingw/x86_64-w64-mingw32/include/crtdefs.h:10:0,
                    from C:/Users/RyanGlScott/Software/ghc-8.0.1/mingw/x86_64-w64-mingw32/include/stddef.h:7,
                    from C:/Users/RyanGlScott/Software/ghc-8.0.1/mingw/lib/gcc/x86_64-w64-mingw32/5.2.0/include/stddef.h:1,
                    from C:\Users\RyanGlScott\Software\ghc-8.0.1/lib/template-hsc.h:4,
                    from dist\build\System\Win32\MinTTY_hsc_make.c:1:
   C:/Users/RyanGlScott/Software/ghc-8.0.1/mingw/x86_64-w64-mingw32/include/_mingw.h:225:0: note: this is the location of the previous definition
    #define _WIN32_WINNT 0x502
    ^
   ```

   I'm not sure if that can be avoided somehow.
4. I added a `isMinTTYHandle` function that takes an arbitrary `HANDLE` as an argument.

I also added `ntdll` to the `extra-libraries` so that we can just link directly against `NtQueryObject`.

Mistuke commented on System/Win32/MinTTY.hsc:
```diff
@@ -0,0 +1,225 @@
+--    \Device\NamedPipe\msys-dd50a72ab4668b33-pty1-to-master
+--
+-- This means we can't rely on "cygwin-" or "msys-" being at the very start,
+-- so we check for their presence using `isInfixOf` instead of `isPrefixOf`.
```
It's a raw device path, so you can just use `takeFileName` to grab just the filename. Then you can just do the prefix match.

RyanGlScott replied:
OK, I've added `filepath` as a build dependency as a result.

Mistuke commented on System/Win32/MinTTY.hsc:
```diff
@@ -0,0 +1,225 @@
+ntQueryObjectNameInformation :: HANDLE -> IO String
+ntQueryObjectNameInformation h = do
+  let sizeOfUSHORT = sizeOf (undefined :: USHORT)
+  let bufSize = 8 * sizeOfUSHORT + mAX_PATH * sizeOfWchar
```
Note that since this seems to return a raw device path, it likely isn't constrained by the ANSI filepath constraints.
That said, It's highly unlikely that a named pipe would be longer than `MAX_PATH` so this is fine.

Mistuke commented on System/Win32/MinTTY.hsc:
```diff
@@ -0,0 +1,225 @@
+instance Storable UNICODE_STRING where
+    sizeOf    _ = #size      UNICODE_STRING
+    alignment _ = #alignment UNICODE_STRING
+    poke buf us = withCWStringLen (usBuffer us) $ \(str, len) -> do
```
can you use `withTStringLen` instead? Just to be consistent.

RyanGlScott replied:
I've replaced all uses of `CWchar` (and auxiliary functions) with `TCHAR`.

Mistuke commented on System/Win32/MinTTY.hsc:
```diff
@@ -0,0 +1,225 @@
+import Control.Exception (catch)
+#endif
+import Data.List (isInfixOf, isSuffixOf)
+import Data.Word (Word8)
```
hmm build seems to be complaining that this import is redundant.

RyanGlScott replied:
Ah, it turns out `Foreign` reexports `Data.Word`. I'll remove the redundant `Data.Word` import then.

Mistuke commented on System/Win32/MinTTY.hsc:
```diff
@@ -0,0 +1,225 @@
+-- The headers that are shipped with GHC's copy of MinGW-w64 assume Windows XP.
+-- Since we need some structs that are only available with Vista or later,
+-- we must manually set WINVER/_WIN32_WINNT accordingly.
+#define WINVER 0x0600
```
You can avoid the warning by adding `#undef _WIN32_WINNT` here.

`hsc2hs` is doing something it shouldn't. It's including it's own header before any of your own, this header includes a system header, which triggers the define of `_WIN32_WINNT`. So it looks like you always have to `#undef` when using `hsc2hs`.

Mistuke commented:
It seems that this no longer builds with older toolchains before 7.10.3.

```
[23:02:23]Warning: The package list for 'hackage.haskell.org' does not exist. Run 'cabal
[23:02:23]update' to download it.
[23:02:40]MinTTY.hsc:51:22: fatal error: winternl.h: No such file or directory
[23:02:40]compilation terminated.
[23:02:40]compiling dist\build\System\Win32\MinTTY_hsc_make.c failed (exit code 1)
[23:02:40]command was: E:\GHC\i386\ghc-7.8.4\mingw\bin\gcc.exe -c dist\build\System\Win32\MinTTY_hsc_make.c -o dist\build\System\Win32\MinTTY_hsc_make.o -U__i686 -fno-stack-protector -D__GLASGOW_HASKELL__=708 -Dmingw32_BUILD_OS=1 -Di386_BUILD_ARCH=1 -Dmingw32_HOST_OS=1 -Di386_HOST_ARCH=1 -Iinclude -fno-strict-aliasing -Idist\build\autogen -include dist\build\autogen\cabal_macros.h -IE:\GHC\i386\ghc-7.8.4\lib\bytestring-0.10.4.0\include -IE:\GHC\i386\ghc-7.8.4\lib\base-4.7.0.2\include -IE:\GHC\i386\ghc-7.8.4\lib\integer-gmp-0.5.1.0\include -IE:\GHC\i386\ghc-7.8.4\lib/include -IE:\GHC\i386\ghc-7.8.4/lib/include/
[23:02:40]Process exited with code 1
```

Because the header isn't there. I would like to keep compat with `7.10.1` and `7.10.2`. Do you need the header only for the function import? If so maybe we can make it just use `GetProcAddress` as well.

RyanGlScott commented:
I certainly would like to keep backwards compatibility too! I use `winternl.h` to get access to the `OBJECT_NAME_INFORMATION` and `UNICODE_STRING` structs (for `isMinTTYCompat`). If there's a better way to bring them into scope, I'd be all for it.

RyanGlScott commented:
And we certainly could just redefine the structs ourselves if need be, if they aren't available via some other header.

Mistuke commented:
Hmm in that case, what do you think about putting those structure in a `winternl-compat.h` file that gets included for `GHC 7.10.2` and earlier instead of `winternl.h`. The structures are unlikely to change, if they do the storable instances will break anyway. I have a similar issue with my own pull requests and I think I'll go this route..

RyanGlScott commented:
Out of curiosity, how did you trigger the build error in https://github.com/haskell/win32/pull/63#issuecomment-267854133? I just tried building this patch with GHC 7.10.2 locally and it seemed to work.

Mistuke commented:
Sorry I completely missed your last message @RyanGlScott .

So my CI builds the Win32 package 6 ways. all og gHC 7.8.4, 7.10.3 and 8.0.1 both 32  and 64 bit.

That error seems to come from the 32 bit version of GHC 7.8.4. The only command I do is a configure and build.

RyanGlScott commented:
Ah, it's an architecture issue. Now that I've re-read [GHC Trac #10726](https://ghc.haskell.org/trac/ghc/ticket/10726), that makes sense - before GHC 7.10.3, we were bundling `MinGW`, not `MinGW-w64`, with the 32-bit version of GHC on Windows, and it turns out that [MinGW never shipped with `winternl.h`](https://sourceforge.net/p/tdm-gcc/bugs/297/). Darn.

I'll work on coming up with a `winternl-compat.h` file and test it on old 32-bit GHCs.

RyanGlScott commented:
I've updated it to have old 32-bit Windows compatibility - feel free to fire off another TeamCity build.
